agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -1146,7 +1146,25 @@ impl DeriveSchema for TaggedOperator {
                     set! {then_schema, else_schema},
                 )))
             }
-            TaggedOperator::Filter(_) => todo!(),
+            TaggedOperator::Filter(f) => {
+                let input_schema = f.input.derive_schema(state)?;
+                let is_nullable = input_schema.satisfies(&NULLISH.clone()) != Satisfaction::Not;
+                let array_schema = array_schema_or_error!(input_schema, f.input);
+                let var_name = f._as.clone().unwrap_or("this".to_string());
+                state.variables.insert(var_name.clone(), array_schema);
+                f.cond.derive_schema(state)?;
+                let filter_schema = state.variables.remove(&var_name);
+                if let Some(schema) = filter_schema {
+                    let mut schema = Schema::Array(Box::new(schema));
+                    if is_nullable {
+                        schema = schema.union(&Schema::Atomic(Atomic::Null))
+                    }
+                    Ok(schema)
+                // this should be unreachable, since we manually add and remove the variable from the state
+                } else {
+                    Ok(Schema::Unsat)
```

Comment:
Should we panic instead of return `Unsat` if it's unreachable? I know Patrick updated an error case to a panic in one of his PRs I reviewed recently. Not sure what our convention is in this situation throughout the library (or if it even matters what choice we make).

Reply:
Yeah this is a good callout. I think the compass I have mostly been following is error when you have a reachable issue -- eg an operator that takes an `Expression` but that expression must be an array literal or field reference but cannot be something else; and panicing when a pipeline is in fact unreachable. I'll do that here!